ARCHITECTURE.md

```diff
@@ -0,0 +1,152 @@
+- Technically the script builds Bluepill & the sample app, and runs the tests
+- `BPSampleApp.app` is the "bundle", an x86_64 bundle but iPhone needs an ARM executable
+	- "Fat binary" will have both
+- Side note: iOS is a variant of MacOS and Android ships a VM with a Linux based Android kernel
+- Originally Apple Emulators translated ARM -> `x86_64` but that was very slow!
+	- Helped test iOS apps in Mac by giving you an `x86_64` executable
+- Now they compile all frameworks for `x86_64` and give you a sandbox in which you can build a native binary
+	- It's faster to get an exectuable to test this way
+
+
+## Some notes on iOS Simulators
+- `xcrun simctl boot` will boot the simulator (Alternatively run Simulator.App in /Applications/Xcode.App)
+- `ps axuw | grep 'launchd_sim'` -> This is the `launchd` process running inside the Simulator
+	- Since the Simulator isn't a VM (like Android) and shares the same Kernel as the underlying Mac OS, it's possible to inspect the processes running inside the simulator
+	- `ps axiw | grep 'CoreSimulator'`
+- `xcrun simctl install .... BPSampleApp.app`
+- Kernel runs with special persmissions (Ring 0) that user programs aren't privy to
+	- Copy of iOS runs in user space of MacOS Kernel
+- The VM model: Multiple Kernels that share the same processor
+	- CPU -> Kernel1 (OS1)/Kernel2 (OS2)/Kernel3 (OS2)
+- `otool -L MacOS/Simulator`
+	- Inspect Frameworks -> /Library/Developer/PrivateFrameworks/CoreSimulator..
+	- Bluepill links with PrivateFrameworks (like simctl)
+- All functionality is provided as a framework. Simluator UI uses this library and so does Bluepill
+- file CoreSimulator
+	- Shows you a binary with 2 architectures for M1 compatibility
+	- CoreSimulator.framework/Versions/A
+		- Poke around to use the framework directly in Bluepill
+
+## iOS Apps Test Classes
+- UI Tests
+	- Tests your app's UI
+	- Uses accessibility API of iOS
+	- Stuff like asserts about things being on screen
+- Unit Tests
+	- Non UI assertions and these things run _very_ differently
+
+## How we rely on XCTest Framework
+`XCTest` is a _Plugin_
+```
+	bluepill/Build ... /BPSampleApp.app
+	tree Base.lproj
+	tree Frameworks
+
+	XCTest.Framework
+	tree Plugins (Used for writing Plugins which is runtime code for extra functionality)
+```
+
+## How Bluepill Executes Tests
+`BPSampleTests.xctest` is a Bundle
+	- Executable
+	- Frameworks
+	- CodeSignature
+	- Info.plist (`plutil -p Info.plist`)
+
+- `FBundleIdentifier` -> Handle that iOS is using for interaction with the App
+- App loads the Plugin (`main -> loadPlugin`)
+- App binary dynamically links to `.dylib` files using the dynamic loader
+	- Dynamic loader maps a `.dylib` hook to its location in memory (Part of OS funcitonality)
+	- `man dyld` (See `DYLD_INSERT_LIBRARIES` option which is what we use in Bluepill)
+	- The _trampoline_ code to do this is in `libSystem.B.dylib` (part of the kernel space)
+- `otool --help`
```

Comment:
I feel we can have these tricks in a separate section or doc, about "useful tools to tinker with stuff"